### DISCUSSION TITLE
StringBuilder or StringFormatter (whatever you want to call it)

### DISCUSSION THREAD

**Discussion by YogiYang** [twinbasic/twinbasic#791] [Ideas]
Give a function called either StringBuilder or StringFormatter for building complex strings using place holders.

It syntax would be something like this:
`<Result> = StringBuilder(<ActualString> [, <Parameter 1> [... [, <Parameter N>]]])`

Here ActualString is a string with place holders for example - `"The age of patient - %1 is %2 and has been diagnosed with %3"`
`Parameter 1` are the actual values that are to be replaced in the ActualString.

Sample usage:
```
varMalady = "Headache"
MyResultString = StringBuilder("The age of patient - %1 is %2 and has been diagnosed with %3", "Yogi Yang", 31, varMalady )
```

Here there is one cache and that is that the number of place holders and parameters should match in number.

**Comment by wqweto** [comment]
This is extremely simple to implement e.g. we are using this `Printf` function in our codebases

```
Public Function Printf(ByVal sText As String, ParamArray A() As Variant) As String
    Const LNG_PRIVATE   As Long = &HE1B6 '-- U+E000 to U+F8FF - Private Use Area (PUA)
    Dim lIdx            As Long
    
    For lIdx = UBound(A) To LBound(A) Step -1
        sText = Replace(sText, "%" & (lIdx - LBound(A) + 1), Replace(A(lIdx), "%", ChrW$(LNG_PRIVATE)))
    Next
    Printf = Replace(sText, ChrW$(LNG_PRIVATE), "%")
End Function
```
The loop is backwards so that `%10` and `%11` do not get messed up by `%1` placeholders. This allows values to contain "%1" strings too that do not get replaced (are not treated as placeholders).

A compiler support will be needed for true string interpolation like in this C# sample
```
// Composite formatting:
Console.WriteLine("Hello, {0}! Today is {1}, it's {2:HH:mm} now.", name, date.DayOfWeek, date);
// String interpolation:
Console.WriteLine($"Hello, {name}! Today is {date.DayOfWeek}, it's {date:HH:mm} now.");
```

**Reply by mwolfe02** [reply]
Hey, @wqweto, that's nifty!  

Would you mind if I wrote an article on that function over at my blog, [nolongerset.com](https://nolongerset.com)?  I would of course credit you and link to your GitHub profile and this GitHub comment.  Let me know if you have other URL(s) that you would prefer I link to (e.g., a company website, your blog, etc.).  Thanks!

**Reply by wqweto** [reply]
@mwolfe02 Sure, no problem!

**Reply by SuperDre** [reply]
String interpolation would be a very nice addition to TwinBasic. Having used it in C# now, I hate not having it in VB6. I'm even redoing my Field replacement texts from 'My Text %SomeVariable% ' to 'My Text {SomeVariable}' just to be more in line with what I'm writing for C#.

**Comment by Greedquest** [comment]
There also some VBA answers here:

https://stackoverflow.com/questions/17233701/obtaining-the-equivalent-to-printf-or-string-format-in-excel

I use
```vba
Public Function printf(ByVal mask As String, ParamArray tokens() As Variant) As String
    Dim i As Long
    For i = 0 To UBound(tokens)
        Dim escapedToken As String
        escapedToken = Replace$(tokens(i), "}", "\}") 'only need to replace closing bracket since {i\} is already invalid
        If InStr(1, mask, "{}") <> 0 Then
            'use positional mode {}
            mask = Replace$(mask, "{}", escapedToken, Count:=1)
         
        Else
            'use indexed mode {i}
            mask = Replace$(mask, "{" & i & "}", escapedToken)
        
        End If
    Next
    mask = Replace$(mask, "\}", "}")
    printf = mask
End Function
```
since it allows you not to number the substitutions (`%1 %2` or `{1} {2}` just become `{} {}`). You escape chars with `\}` but that said, I quite like @wqweto 's backwards looping thing

**Comment by cristianbuse** [comment]
Doing substitutions covers a lot of cases but is never clean. For example in the solution provided by @wqweto if we try something like:

```VBA
Printf("Hello %1 or %%2", "John", "1 Clare")
```
then we get ```Hello John or John Clare``` which is wrong. It should be ```Hello John or %1 Clare```.

Same with the solution provided by @Greedquest , and as it can be seen in the link he provided (as a note) something like ```{0\}{1}``` in the mask  returns the wrong result. Or a more concrete example like ```Printf("Hello {{0}} > {1}", "1", "Test")``` which returns ```Hello Test > Test``` instead of ```Hello {1} > Test```.

The only way to be sure there are no unexpected edge cases is to parse the mask. It can be done character by character or as a split.

For example, updating the '%' solution to be safe using a split approach could look like:

```VBA
Public Function Printf(ByVal mask As String, ParamArray tokens() As Variant) As String
    Dim parts() As String: parts = Split(mask, "%")
    Dim i As Long
    Dim j As Long
    Dim isFound As Boolean
    Dim s As String
    '
    'Always ignore first part - covers if mask started or not with %
    For i = LBound(parts) + 1 To UBound(parts)
        If LenB(parts(i)) = 0 Then
            parts(i) = "%"
        Else
            isFound = False
            For j = UBound(tokens) To LBound(tokens) Step -1
                s = CStr(j + 1)
                If Left$(parts(i), Len(s)) = s Then
                    parts(i) = tokens(j) & Right$(parts(i), Len(parts(i)) - Len(s))
                    isFound = True
                    Exit For
                End If
            Next j
            If Not isFound Then
                parts(i) = "%" & parts(i)
            End If
        End If
    Next i
    Printf = Join(parts, vbNullString)
End Function
```

which is more efficient if the tokens have a lot of % characters or slower if the mask has a lot of % characters but most importantly is correct.

**Reply by WaynePhillipsEA** [reply]
Good to see you're still around Cristian 👍

**Reply by mwolfe02** [reply]
@cristianbuse I just re-posted your function in [an article I wrote](https://nolongerset.com/printf-function/) about @wqweto's function.  Please let me know if you would rather I remove the function from my article and replace it with a link to this page.

**Reply by wqweto** [reply]
Somehow didn't got notified about this comment.

> The only way to be sure there are no unexpected edge cases is to parse the mask. It can be done character by character or as a split.

Btw, here is another one with replacements which uses the already deemed `ChrW(&HE1B6)` as non-usable in format string.
```
Public Function Printf(ByVal sText As String, ParamArray A() As Variant) As String
    Const LNG_PRIVATE   As Long = &HE1B6 '-- U+E000 to U+F8FF - Private Use Area (PUA)
    Dim lIdx            As Long
    
    With CreateObject("VBScript.RegExp")
        .Pattern = "%(\d+)"
        .Global = True
        sText = .Replace(sText, ChrW$(LNG_PRIVATE) & "$1")
    End With
    For lIdx = UBound(A) To LBound(A) Step -1
        sText = Replace(sText, ChrW$(LNG_PRIVATE) & (lIdx - LBound(A) + 1), A(lIdx))
    Next
    Printf = sText
End Function
```
Not tested very much but works on the failing edge-case.

**Reply by cristianbuse** [reply]
@mwolfe02 I don't mind you re-posting anything that I post. Appreciate you asking but no need to ask going forward.

**Reply by cristianbuse** [reply]
@wqweto 

I appreciate that the 'special' character would be a rare case but this code would be perfectly legal:

```VBA
Printf("Hello %1 or %%2", "John", ChrW$(&HE1B6) & "1 Clare")
```
and it would still return the wrong result: ```Hello John or %John Clare```.

Moreover, regEx is just too slow to be used in a library.

As others commented, this kind of function needs to sit in a library because it would be heavily used anyway by everyone. My take on this is that the library function needs to be 'clean' meaning no tricks like the one you presented. I appreciate the cleverness of the solution but it's still a trick in my view. I only provided a quick alternative which I created just to prove the point but I don't like my solution either. I'm certain that Wayne can come up with a clean and fast parsing solution to this.

**Reply by wqweto** [reply]
Completely agree a clean and safe implementation is better. Will test performance on 3 impl now.

Edit: Using the failing edge-case above:
```
Printf1 in 0.271
Printf2 in 0.187
Printf3 in 0.276
```
Apparently your impl is faster too as it does not allocate BSTRs in a loop. Great job!

**Reply by FullValueRider** [reply]
Could I make a request that string interpolation also includes the ability to stringify and print arrays, 'collection' objects and some form of other representation if an item cannot be stringified.  I made a crude attempt at this with the 'Fmt' class which is included in the TwinLib64 package I uploaded to the package manager.

**Comment by YogiYang** [comment]
Thanks all for your replies and alternative solutions.

Instead of debating so much why not request @WaynePhillipsEA to add a function in the compiler to achieve this in a robust manner.

One cannot expect a fresh graduate out of school (or a newbie to tB) to be able to build such a complex function so it should be available by default as a builtin function.

Actually this and many such functions and procedures if provided out-of-the-box in tB will draw new developers towards it.

**Reply by Greedquest** [reply]
Oh yeah I definitely want string interpolation to be part of the language (not necessarily in the compiler, perhaps as an overload of the String type or just a standard library or even built-in function)

**Comment by FullValueRider** [comment]
I'm not seeing significant value in having the compiler do string interpolation.  Substitution of a field marker in a string by a variable can readily be achieved through a library.  I know some languages now allow substitution by variable name, but I think in the long run this adds less value than you think, particularly when you want to start formatting the appearance of an interpolated variable.

So my vote, if we are voting, is for string interpolation to be a library which can be written and policed by us users, rather than a diversion from getting a good compiler.

**Reply by EduardoVB** [reply]
> So my vote, if we are voting, is for string interpolation to be a library which can be written and policed by us users, rather than a diversion from getting a good compiler.

At some point I suggested to have some file functions (#102), and the consensus was that they would be better implemented by libraries. I mention it because I think it is a very similar case.
